source/common/stats/stats_impl.cc

```diff
@@ -122,6 +122,66 @@ RawStatData* HeapRawStatDataAllocator::alloc(const std::string& name) {
   return data;
 }
 
+TagProducerImpl::TagProducerImpl(const envoy::api::v2::StatsConfig& config) : TagProducerImpl() {
+  // To check name conflict.
+  std::unordered_set<std::string> names;
+  reserveResources(config);
+  addDefaultExtractors(config, names);
+
+  for (const auto& tag_specifier : config.stats_tags()) {
+    if (!names.emplace(tag_specifier.tag_name()).second) {
+      throw EnvoyException(fmt::format("Tag name '{}' specified twice.", tag_specifier.tag_name()));
+    }
+
+    // If no tag value are found, fallback to default regex to keep backward compatibility.
+    if (tag_specifier.tag_value_case() == envoy::api::v2::TagSpecifier::TAG_VALUE_NOT_SET ||
+        tag_specifier.tag_value_case() == envoy::api::v2::TagSpecifier::kRegex) {
+      tag_extractors_.emplace_back(Stats::TagExtractorImpl::createTagExtractor(
+          tag_specifier.tag_name(), tag_specifier.regex()));
+
+    } else if (tag_specifier.tag_value_case() == envoy::api::v2::TagSpecifier::kFixedValue) {
+      default_tags_.emplace_back(Stats::Tag{tag_specifier.tag_name(), tag_specifier.fixed_value()});
+    }
+  }
+}
+
+TagProducerImpl::TagProducerImpl() {}
```

Comment:
nit: del empty constructor and destructor (if default constructor is needed for some reason (?) just define in header file)